File: behavioral/observer/observer.py

# Subject interface
class Subject:
    def attach(self, observer):
        pass

    def detach(self, observer):
        pass

    def notify(self):
        pass

# Concrete Subject
class ConcreteSubject(Subject):
    def __init__(self):
        self._observers = []
        self._state = None

    def get_state(self):
        return self._state

    def set_state(self, state):
        self._state = state
        self.notify()

    def attach(self, observer):
        self._observers.append(observer)

    def detach(self, observer):
        self._observers.remove(observer)

    def notify(self):
        for observer in self._observers:
            observer.update()

# Observer interface
class Observer:
    def update(self):
        pass

# Concrete Observer
class ConcreteObserver(Observer):
    def __init__(self, subject):
        self._subject = subject
        self._subject.attach(self)

    def update(self):
        print(f"Observer updated with state: {self._subject.get_state()}")

if __name__ == "__main__":
    subject = ConcreteSubject()

    observer1 = ConcreteObserver(subject)
    observer2 = ConcreteObserver(subject)

    subject.set_state("State1")
    subject.set_state("State2")

File: src/behavioral/observer/Observer.cs
namespace DesignPatterns.Behavioral.Observer;

// Subject interface
public interface ISubject
{
    void Attach(IObserver observer);
    void Detach(IObserver observer);
    void Notify();
}

// Concrete Subject
public class ConcreteSubject : ISubject
{
    private List<IObserver> _observers = new();
    private string _state;

    public void SetState(string state)
    {
        _state = state;
        Notify();  // Notifies observers when state changes
    }

    public void Attach(IObserver observer)
    {
        _observers.Add(observer);  // Adds observer to the list
    }

    public void Detach(IObserver observer)
    {
        _observers.Remove(observer);  // Removes observer from the list
    }

    public void Notify()
    {
        foreach (var observer in _observers)
        {
            observer.Update();  // Notifies all observers
        }
    }

    public string GetState() => _state;  // Returns current state
}

// Observer interface
public interface IObserver
{
    void Update();  // Method to update the observer
}

// Concrete Observer
public class ConcreteObserver : IObserver
{
    private ConcreteSubject _subject;

    public ConcreteObserver(ConcreteSubject subject)
    {
        _subject = subject;
        _subject.Attach(this);  // Automatically attach to subject
    }

    public void Update()
    {
        Console.WriteLine("Observer updated with state: " + _subject.GetState());
    }
}

// Client code
class Program
{
    static void Main()
    {
        var subject = new ConcreteSubject();

        var observer1 = new ConcreteObserver(subject);
        var observer2 = new ConcreteObserver(subject);

        subject.SetState("State1");
        subject.SetState("State2");
    }
}
